File: Problem2_Isomorphic.cpp
// Time Complexity : O(n) n is string length
// Space Complexity : O(n*n) as we are using two hashmaps
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : no


// Your code here along with comments explaining your approach


class Solution {
public:
    bool isIsomorphic(string s, string t) {
        
        if(s.length() != t.length()) return false;
        
        unordered_map<char, char> sMap;
        unordered_map<char, char> tMap;
        
        for(int i=0; i<s.length();i++){
            char sChar=s.at(i);
            char tChar=t.at(i);
            
            if(sMap.count(sChar)){
                
                //unordered_map<char, char>::iterator it= sMap.find(sChar);
                auto it= sMap.find(sChar);
                
                //cout<<it->first<<endl;
                if(it->second != tChar) return false;
            }else {
                sMap.insert({sChar,tChar});
            }
            if(tMap.count(tChar)){
                
                //unordered_map<char, char>::iterator it= sMap.find(sChar);
                auto it= tMap.find(tChar);
                
                //cout<<it->first<<endl;
                if(it->second != sChar) return false;
            }else {
                tMap.insert({tChar,sChar});
            }
        }
        
                
        return true;
    }
};

File: problem2_isomorphic.py
#time complexity : O(n) when n is the number of input string 
#space complexity: O(n), because here two dictionaries (mapS and mapT) are used to store the mapping of characters from string s to t and t to s
#Submit on leetcode : yes

class Solution:
    def isIsomorphic(self, s: str, t: str) -> bool:
        if len(s) != len(t): 
            return False
        if not s and not t:
            return False

        mapS, mapT = {}, {}

        for i in range(len(s)):
            charS, charT = s[i], t[i]

            if (charS in mapS and mapS[charS] != charT) or (charT in mapT and mapT[charT] != charS):
                return False

            mapS[charS] = charT
            mapT[charT] = charS

        return True


solution = Solution() 
s1 = "egg"
t1 = "add"
print(f"Isomorphic for '{s1}' and '{t1}': {solution.isIsomorphic(s1, t1)}")      
        